File: code/tree.hs

type Gift = String

data ChristmasTree a = Nil
  | Leaf a
  | Branch (ChristmasTree a) (ChristmasTree a)
  deriving (Eq, Ord, Read)


-- TODO: write down something to find all the gifts in a given tree

find_gifts :: ChristmasTree Gift -> [Gift]
find_gifts Nil = []
find_gifts (Leaf gift) = [gift]
find_gifts (Branch left right) = (find_gifts left) ++ (find_gifts right)

instance (Show a) => Show (ChristmasTree a) where
  show Nil = "O"
  show (Leaf gift) = "G"
  show (Branch left right) = "*\n" ++ (show left) ++ " ~~ " ++ (show right)
  
main :: IO ()
main = do
  print $ Branch (Leaf "Toy") (Branch (Leaf "Ball") (Leaf "Toy"))

File: code/tree.py
"""
Simple Tree representation in Python
"""


class Tree:
    def __init__(self):
        pass

    def find_gifts(self):
        raise NotImplementedError

    @classmethod
    def parse(cls, tree_string):
        pass


class Branch(Tree):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def find_gifts(self):
        return self.left.find_gifts() + self.right.find_gifts()


class Leaf(Tree):
    def __init__(self, gift=None):
        self.gift = gift

    def __str__(self):
        return "Nothing" if self.gift is None else self.gift

    def find_gifts(self):
        return [self.gift]


def main():
    mytree = Branch(Leaf(gift="Toy"), Leaf())
    print(mytree.find_gifts())


if __name__ == '__main__':
    main()
